src/Microsoft.AspNetCore.Mvc.Core/Formatters/TextInputFormatter.cs

```diff
@@ -0,0 +1,97 @@
+
+        /// <summary>
+        /// Gets the mutable collection of character encodings supported by
+        /// this <see cref="TextInputFormatter"/>. The encodings are
+        /// used when reading the data.
+        /// </summary>
+        public IList<Encoding> SupportedEncodings { get; } = new List<Encoding>();
+
+        /// <inheritdoc />
+        public override Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterContext context)
+        {
+            var request = context.HttpContext.Request;
+            var selectedEncoding = SelectCharacterEncoding(context);
+            if (selectedEncoding == null)
+            {
+                var message = Resources.FormatUnsupportedContentType(
+                    context.HttpContext.Request.ContentType);
+
+                var exception = new UnsupportedContentTypeException(message);
+                context.ModelState.AddModelError(context.ModelName, exception, context.Metadata);
+
+                return InputFormatterResult.FailureAsync();
+            }
+
+            return ReadRequestBodyAsync(context, selectedEncoding);
+        }
+
+        /// <summary>
+        /// Reads an object from the request body.
+        /// </summary>
+        /// <param name="context">The <see cref="InputFormatterContext"/>.</param>
+        /// <param name="encoding">The <see cref="Encoding"/> used to read the request body.</param>
+        /// <returns>A <see cref="Task"/> that on completion deserializes the request body.</returns>
+        public abstract Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding);
+
+        /// <summary>
+        /// Returns an <see cref="Encoding"/> based on <paramref name="context"/>'s
+        /// character set.
+        /// </summary>
+        /// <param name="context">The <see cref="InputFormatterContext"/>.</param>
+        /// <returns>
+        /// An <see cref="Encoding"/> based on <paramref name="context"/>'s
+        /// character set. <c>null</c> if no supported encoding was found.
+        /// </returns>
+        protected Encoding SelectCharacterEncoding(InputFormatterContext context)
+        {
+            var request = context.HttpContext.Request;
+
+            var requestEncoding = request.ContentType == null ? null : MediaType.GetEncoding(request.ContentType);
+            if (requestEncoding != null)
+            {
+                for (int i = 0; i < SupportedEncodings.Count; i++)
+                {
+                    if (string.Equals(
+                        requestEncoding.WebName,
+                        SupportedEncodings[i].WebName,
+                        StringComparison.OrdinalIgnoreCase))
+                    {
+                        return SupportedEncodings[i];
+                    }
+                }
+
+                return null;
+            }
+
+            return SupportedEncodings.Count > 0 ? SupportedEncodings[0] : null;
```

Comment:
Leave some comments here explaining **WHY** we want L91 to return null and L94 to do a first-or-default